### PR TITLE
Issue 2041

### PR DESCRIPTION
**Pull request by Nastsi**
Deserialization JSON from body of Post-request

### CODE REVIEW

File: RedirectApplication/RedirectApplication/Controllers/ValuesController.cs [hunk 1]
```diff
@@ -37,10 +40,13 @@ public string Get(int id)
         {
             return "value";
         }
-
+
```
Comment (102) [inline RedirectApplication/RedirectApplication/Controllers/ValuesController.cs]: remove trailing spaces

File: RedirectApplication/RedirectApplication/Json/Json.cs [hunk 2]
```diff
@@ -16,12 +16,20 @@ public override void WriteJson(JsonWriter writer, object value, JsonSerializer s
...
-            return array.Children().Select(child => {
-                var node = CreateNode(child);
-                serializer.Populate(child.CreateReader(), node);
-                return node;
-            }).ToList();
+            if (reader.Path == "TargetUrl")
+            {
+                return reader.Value;
+            }
+            else
```
Comment (mvahrameev) [inline RedirectApplication/RedirectApplication/Json/Json.cs]: `else` is redundant

File: RedirectApplication/RedirectApplication/Controllers/ValuesController.cs [hunk 3]
```diff
@@ -39,8 +42,11 @@ public string Get(int id)
         }
 
         // POST api/values
-        public void Post([FromBody]string value)
+        public void Post(HttpRequestMessage request)
         {
+            var someText = request.Content.ReadAsStringAsync().Result;
+            var reader = new JsonTextReader(new StringReader(someText));
+            PostJson content = JsonSerializer.CreateDefault().Deserialize<PostJson>(reader);
```
Comment (maxmalov) [inline RedirectApplication/RedirectApplication/Controllers/ValuesController.cs:54]: This method isn't doing anything actually

File: RedirectApplication/RedirectApplication/Json/Json.cs [hunk 4]
```diff
@@ -16,8 +16,13 @@ public override void WriteJson(JsonWriter writer, object value, JsonSerializer s
 
         public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
         {
+            if (reader.Path == "TargetUrl")
```
Comment (maxmalov) [inline RedirectApplication/RedirectApplication/Json/Json.cs:19]: Looks like a hack. Can you elaborate a little?

File: RedirectApplication/RedirectApplication/Controllers/ValuesController.cs (hunk 3 above)
Comment (Nastsi) [reply RedirectApplication/RedirectApplication/Controllers/ValuesController.cs:54]: In this method I get JSON from Post-request and deserialize it in PostJson object

Comment (maxmalov) [reply RedirectApplication/RedirectApplication/Controllers/ValuesController.cs:54]: aaaaand then...?

**Comment by maxmalov** [comment]
As was sad earlier: https://github.com/hwdtech/redirect/pull/7#issuecomment-255125749

File: RedirectApplication/RedirectApplication/Controllers/ValuesController.cs (hunk 3 above)
Comment (Nastsi) [reply RedirectApplication/RedirectApplication/Controllers/ValuesController.cs:54]: What happened then will be written in another tickets. This one is about getting JSON and deserialization

File: RedirectApplication/RedirectApplication/Json/Json.cs (hunk 4 above)
Comment (Nastsi) [reply RedirectApplication/RedirectApplication/Json/Json.cs:19]: Example of JSON:

{
"TargetUrl":"localhost",
"Conditions": [ {"Name:":"ByBrowser","Browser":"Chrome","Url":"http://pikabu.ru/"} ]
}

Either we have just one field "TargetUrl", either we working with everything in "Conditions" (that may look like a tree).
I just don't know how to complicate this

File: RedirectApplication/RedirectApplication/Controllers/ValuesController.cs (hunk 3 above)
Comment (maxmalov) [reply RedirectApplication/RedirectApplication/Controllers/ValuesController.cs:54]: So again, it should be in another tickets PRs https://github.com/hwdtech/redirect/pull/7#discussion_r84063406

File: RedirectApplication/RedirectApplication/Controllers/ValuesController.cs [hunk 5]
```diff
@@ -39,8 +42,16 @@ public string Get(int id)
         }
 
         // POST api/values
-        public void Post([FromBody]string value)
+        public void Post(HttpRequestMessage request)
         {
+            Deserialization(request);
+        }
+
+        public void Deserialization(HttpRequestMessage request)
```
Comment (maxmalov) [inline RedirectApplication/RedirectApplication/Controllers/ValuesController.cs]: I was talking about such schema:

``` csharp
public void Post(...)
{
  // var content = DeserializeRecord(...);
}

private PostJson DeserializeRecord(HttpRequestMessage request)
{
  // return deserialized record
}
```

Comment (Nastsi) [reply RedirectApplication/RedirectApplication/Controllers/ValuesController.cs]: I did this in the first place but then rethink

File: RedirectApplication/RedirectApplication/Json/Json.cs (hunk 4 above)
Comment (102) [reply RedirectApplication/RedirectApplication/Json/Json.cs:19]: @Nastsi Use markdown syntax to insert code examples into comments, i.e.:

``` json
{
  "TargetUrl":"localhost",
  "Conditions": [ {"Name:":"ByBrowser","Browser":"Chrome","Url":"http://pikabu.ru/"} ]
}
```

will be generated from

``````
```json
{
  "TargetUrl":"localhost",
  "Conditions": [ {"Name:":"ByBrowser","Browser":"Chrome","Url":"http://pikabu.ru/"} ]
}
``` 
``````

**Review by jacob-l — APPROVED** [review]
We can merge it now. Later I will ask to change several things

**Pull request merged**